### PR TITLE
refactor(polys): unify gmpy and non-gmpy domains

### PR DESCRIPTION
**Pull request by oscarbenjamin**
Move gmpy handling code from sympy.core.compatibility to
sympy.external.gmpy. Move PythonRational to PythonMPQ in
sympy.external.pythonmpq.

It is now possible to do

  from sympy.external.gmpy import MPQ

which will either give gmpy2's mpq type or PythonMPQ.

Future work could remove the alternate versions of ZZ and QQ.

#### Release Notes

* polys
    * The polynomials module now only supports using gmpy2 as an accelerator for the `ZZ` and `QQ` domains (gmpy 1.x is no longer supported).

### CODE REVIEW

File: sympy/compatibility/pythonmpq.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    # The hashing algorithm for Fraction changed in Python 3.8
+
+    if sys.version_info >= (3, 8):
+        #
+        # Hash for Python 3.8 onwards
+        #
+        def __hash__(self):
+            """hash - same as mpq/Fraction"""
+            try:
+                dinv = pow(self.denominator, -1, _PyHASH_MODULUS)
```
Comment (jksuom) [inline sympy/compatibility/pythonmpq.py:144]: It seems that negative exponents were introduced only in 3.8. The exponent used to be `_PyHASH_MODULUS - 2`.

File: sympy/compatibility/pythonmpq.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            x2 = gcd(bq, aq)
+            p, q = ((ap//x1)*(bq//x2), (aq//x2)*(bp//x1))
+        elif isinstance(other, int):
+            x = gcd(other, self.numerator)
+            p = self.numerator//x
+            q = self.denominator*(other//x)
+        else:
+            return NotImplemented
+
+        return self._new_check(p, q)
```
Comment (jksuom) [inline sympy/compatibility/pythonmpq.py:342]: Will a check be necessary?

File: sympy/compatibility/pythonmpq.py (hunk 1 above)
Comment (jksuom) [reply sympy/compatibility/pythonmpq.py:144]: I believe that the old exponent could be used as long as 3.6 and 3.7 are supported.  The internal process will probably be the same.

Comment (jksuom) [reply sympy/compatibility/pythonmpq.py:144]: I was mistaken. The new implementation with exponent -1 is faster. So it is reasonable to implement both.

Comment (oscarbenjamin) [reply sympy/compatibility/pythonmpq.py:144]: We can just remove the old implementation when support for Python 3.7 is dropped but if the hash changes again in future then we might need a version check here again.

File: sympy/compatibility/pythonmpq.py (hunk 2 above)
Comment (oscarbenjamin) [reply sympy/compatibility/pythonmpq.py:342]: It's possible that `q == 0` e.g. if other is `int(0)`. It's also possible that `q < 0`. The check is to handle those cases so it is necessary.

Comment (jksuom) [reply sympy/compatibility/pythonmpq.py:342]: Ah..that is true even though p and q are coprime. I think that in the old PythonRational code it was possible to indicate that a gcd computation could be skipped but maybe that would not save much time here.

Comment (oscarbenjamin) [reply sympy/compatibility/pythonmpq.py:342]: Using `_new_check` still skips the gcd. There are three levels:
```
1. __new__ cancels gcd and calls _new_check
2. _new_check handles q <= 0 and calls _new
3. _new simply creates the object
```

**Comment by oscarbenjamin** [comment]
I intend to do more work on this but I'd like to get this part merged first. The diff here is already big and I already had to fix merge conflicts a few times.

What this does is:

1. Move PythonRational to PythonMPQ in sympy.external
2. Add a sympy.external.gmpy module which exports MPQ that can be either PythonMPQ or gmpy2.mpq
3. Drops support for gmpy (rather than gmpy2)
4. Unifies the ZZ, FF and QQ domains so that whether or not gmpy2 is installed the same class is always used for the domain (although a different class is used for the domain elements).
5. Fixes `Rational.numerator` to be an attribute rather than a method as specified in the numeric tower.
6. Fixes some doc rendering problems in some polys modules where latex is used instead of code format.

Following up from this I will send a PR(s) that:

1. Deletes the ZZ_gmpy etc domain code.
2. Moves all handling ground types to sympy.external
3. Makes Rational use MPQ internally.
4. Moves all the code that imports from gmpy2 or is conditional on gmpy2 to the sympy.external module.

**Comment by oscarbenjamin** [comment]
Does anyone want to review this?

**Comment by jksuom** [comment]
This looks good to me. I was only waiting to see if you wanted to continue.

**Comment by oscarbenjamin** [comment]
Thanks!

**Comment by MarkCBell** [comment]
This PR also including changing the SymPy Rational number class to conform with the [Python `numbers.Rational` in the numeric tower](https://docs.python.org/3/library/numbers.html) by [making their numerator and denominator properties](https://github.com/sympy/sympy/pull/21016/commits/1bb6592822789e0a047619a151724a47a569d8f9) (they were previously methods). While increasing compatibility is great, it might have been good to include this change in the 1.9 release notes since this means that code like
```
import sympy as sp
x = sp.Rational(3, 5)
y = x.numerator() + y.denominator()
```
which worked fine in SymPy < 1.9, now raises `TypeError: 'int' object is not callable` since it is trying to evaluate `y = 3() + 5()`.

**Comment by MarkCBell** [comment]
So it looks like the numerator and denominator methods were added to [SymPy 1.5 in April 2019](https://github.com/sympy/sympy/commit/7d3e6e7516f2780ede5dcf4f0879040a4c6a77f5#diff-b74e859948738d468a39752a95aca90c748c22ab17fffb98f683845c33d3b218R1863-R1868). But using the `.p` and `.q` attributes is a really good suggestion and completely deals with the compatibility issue, thanks!

**Pull request merged**